internal/local/archive.go

```diff
@@ -19,143 +16,77 @@ const (
-type ArchiveReader interface {
-	Next() (FileHeader, error)
-	Read(b []byte) (int, error)
-}
-
-// FileHeader holds the path and file information
-type FileHeader struct {
-	Path string
-	Info os.FileInfo
-}
-
-func newArchiveReader(path string, file *os.File) (ArchiveReader, error) {
-	ext := strings.ToLower(filepath.Ext(path))
-
-	if ext == extZip {
-		return newZipReader(file)
-	}
-
-	if ext == extTar {
-		return newTarReader(file), nil
-	}
-
-	if ext == extTgz || strings.HasSuffix(path, extTarGz) {
-		return newGzipReader(file)
-	}
-
-	if ext == "" {
-		return newDirReader(path)
-	}
+type archiveMatcher func(path, ext string) bool
 
-	return nil, errUnknownArchiveExtension(path)
+func isZip(_, ext string) bool {
+	return ext == extZip
 }
 
-func errUnknownArchiveExtension(path string) error {
-	return fmt.Errorf(
-		"failed to read archive file at %s: unsupported format, use one of [%s] instead",
-		path,
-		strings.Join(supportedExts, ", "),
-	)
+func isTar(_, ext string) bool {
+	return ext == extTar
 }
 
-type zipReader struct {
-	*zip.Reader
-
-	currFileIdx  int
-	currOpenFile io.ReadCloser
+func isTarCompressed(path, ext string) bool {
+	return ext == extTgz || strings.HasSuffix(path, extTarGz)
 }
 
-func newZipReader(f *os.File) (ArchiveReader, error) {
-	fi, err := f.Stat()
-	if err != nil {
-		return nil, err
-	}
-
-	zr, err := zip.NewReader(f, fi.Size())
-	if err != nil {
-		return nil, err
+var (
+	preferredArchiveMatchers = []archiveMatcher{
+		isZip,
+		isTar,
+		isTarCompressed,
 	}
+)
 
-	return &zipReader{Reader: zr, currFileIdx: -1}, nil
+type archive struct {
+	path  string
+	isDir bool
 }
 
-func (r *zipReader) Next() (FileHeader, error) {
-	if r.currOpenFile != nil {
-		r.currOpenFile.Close()
-		r.currOpenFile = nil
-	}
-
-	r.currFileIdx++
-	if r.currFileIdx >= len(r.Reader.File) {
-		return FileHeader{}, io.EOF
-	}
+func selectArchive(archives []string) (archive, error) {
+	extCache := map[string]string{}
 
-	f := r.Reader.File[r.currFileIdx]
+	for _, archiveMatcher := range preferredArchiveMatchers {
+		for _, path := range archives {
+			ext := strings.ToLower(filepath.Ext(path))
 
-	return FileHeader{f.Name, f.FileInfo()}, nil
-}
-
-func (r *zipReader) Read(b []byte) (int, error) {
-	if r.currFileIdx < 0 || r.currFileIdx >= len(r.Reader.File) {
-		return 0, io.EOF
-	}
+			if archiveMatcher(path, ext) {
+				return archive{path: path}, nil
+			}
 
-	if r.currOpenFile == nil {
-		f, err := r.Reader.File[r.currFileIdx].Open()
-		if err != nil {
-			return 0, err
+			extCache[path] = ext
 		}
-		r.currOpenFile = f
 	}
 
-	return r.currOpenFile.Read(b)
-}
-
-type tarReader struct {
-	*tar.Reader
-}
-
-func newTarReader(r io.Reader) ArchiveReader {
-	return &tarReader{Reader: tar.NewReader(r)}
-}
-
-func (r *tarReader) Next() (FileHeader, error) {
-	h, err := r.Reader.Next()
-	if err != nil {
-		return FileHeader{}, err
+	for _, path := range archives {
+		ext, ok := extCache[path]
```

Comment:
the purpose for it was to avoid calling `strings.ToLower(filepath.Ext(path))` twice on each path... probably an unnecessary optimization in the grand scheme of cli-things but here we are /shrug